File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        #dfs
        
        res = []
        if not digits or len(digits) == 0:
            return res
        dic = {'2': 'abc', '3': 'def', '4': 'ghi', '5':'jkl', '6': 'mno', '7':'pqrs', '8':'tuv', '9':'wxyz'}
        def backtrack(idx, lst):
            nonlocal res, dic
            if idx == len(digits):
                s = ''.join(lst)
                res.append(s)
                return
            for c in dic[digits[idx]]:
                lst.append(c)
                backtrack(idx + 1, lst)
                lst.pop()
        backtrack(0, [])
        return res
            

File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.java
class Solution {
    String[] phone = new String[]{"", "", "abc", "def",
                                 "ghi", "jkl", "mno",
                                 "pqrs", "tuv", "wxyz"};
    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        if(digits == null || digits.length() == 0) return res;
        String path = "";
        dfs(digits, 0, res, path);
        return res;
    }
    public void dfs(String digits, int idx, List<String> res, String path) {
        if(idx == digits.length()) {
            res.add(path);
            return;
        }
        for(char c : phone[digits.charAt(idx) - '0'].toCharArray()) {
            dfs(digits, idx + 1, res, path + c);
        }
    }
}